compiler/rustc_expand/src/mbe/metavar_expr.rs

```diff
@@ -0,0 +1,159 @@
+    Ignore(Ident),
+
+    /// The index of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition. The `usize` is the depth.
+    Index(usize),
+
+    /// The length of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition. The `usize` is the depth.
+    Length(usize),
+}
+
+impl MetaVarExpr {
+    /// Attempt to parse a meta-variable expression from a token stream.
+    crate fn parse<'sess>(
+        input: &TokenStream,
+        sess: &'sess ParseSess,
+    ) -> PResult<'sess, MetaVarExpr> {
+        let mut tts = input.trees();
+        match tts.next() {
+            Some(TokenTree::Token(token)) if let Some((ident, false)) = token.ident() => {
+                let Some(TokenTree::Delimited(_, token::Paren, args)) = tts.next() else {
+                    let msg = "meta-variable expression parameter must be wrapped in parentheses";
+                    return Err(sess.span_diagnostic.struct_span_err(ident.span, msg));
+                };
+                let mut iter = args.trees();
+                let rslt = match &*ident.as_str() {
+                    "count" => parse_count(&mut iter, sess, ident.span)?,
+                    "ignore" => MetaVarExpr::Ignore(parse_ident(&mut iter, sess, ident.span)?),
+                    "index" => MetaVarExpr::Index(parse_depth(&mut iter, sess, ident.span)?),
+                    "length" => MetaVarExpr::Length(parse_depth(&mut iter, sess, ident.span)?),
+                    _ => {
+                        let err_msg = "unrecognized meta-variable expression";
+                        let mut err = sess.span_diagnostic.struct_span_err(ident.span, err_msg);
+                        err.span_suggestion(
+                            token.span,
+                            "supported expressions are count, ignore, index and length",
+                            String::new(),
+                            Applicability::MachineApplicable,
+                        );
+                        return Err(err);
+                    }
+                };
+                Ok(rslt)
+            }
+            _ => return Err(sess.span_diagnostic.struct_err("expected meta-variable expression"))
+        }
+    }
+
+    crate fn ident(&self) -> Option<&Ident> {
+        match self {
+            MetaVarExpr::Count(ident, _) | MetaVarExpr::Ignore(ident) => Some(&ident),
+            MetaVarExpr::Index(..) | MetaVarExpr::Length(..) => None,
+        }
+    }
+}
+
+/// Tries to convert a literal to an arbitrary type
+fn convert_literal<T>(lit: Lit, sess: &ParseSess, span: Span) -> PResult<'_, T>
+where
+    T: FromStr,
+    <T as FromStr>::Err: Display,
+{
+    if lit.suffix.is_some() {
+        let msg = "literal suffixes are not supported in meta-variable expressions";
+        return Err(sess.span_diagnostic.struct_span_err(span, msg));
+    }
+    lit.symbol.as_str().parse::<T>().map_err(|e| {
```

Comment:
This will accept any literals parsing to integers, not integer literals, and also won't perform any escaping etc.
You only need to use something like `LitKind::from_lit_token` to extract the semantic literal here.

Reply:
Also it's probably better to inline this into `parse_depth`, all the restrictions here (like no suffixes) are depth-specific.